feat: use langchain abstraction

juftin opened the pull request:
I've been trying to get Elia to work with different LLMs and I figured LangChain would be the best abstraction to use for this. This PR uses the LangChain `BaseChatModel` abstraction for the current models instead of the openapi SDK. 

This should be make this easy to get to work with models like Llama2.

Let me know what you think, happy to adapt this if it's something you're interested in.

darrenburns commented:
I love it - this is definitely really interesting and I'd love to see elia be used with other models! If I knew about this it's definitely the path I'd have taken 😂 

Definitely open to continuing with this.

juftin commented:
Alright, I'm considering this work mostly dev complete. I've gone ahead and replaced the `ChatMessage` TypedDict for a LangChain `BaseMessage`. This is a pydantic model that could be one of `UserMessage`, `AIMessage`, `SystemMessage`. 

Using LangChain I've also implemented a fix to #9 where messages are pruned when they overflow a model's token limit. 

The actual LangChain implementation is here: 
```python
selected_model: GPTModel = MODEL_MAPPING[self.chat_data.model_name]
llm: BaseChatModel = selected_model.model
trimmed_messages = self.trim_messages(
    model=llm,
    messages=self.chat_data.messages,
    max_tokens=selected_model.token_limit,
    preserve_system_message=True,
)
streaming_response = llm.astream(input=trimmed_messages)
```

I don't love how I'm grabbing the `selected_model` out of the `MODEL_MAPPING` dict - but it worked nicely using the chat data's model name a an indexer.

darrenburns reviewed (approved):
Looks great, thank you!
